### DISCUSSION TITLE
Avalonia overwriting existing DataContext from ctor When none provided in axaml

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#12843] [Q&A]
Following the discussion from #12790, Let's say someone doesn't want to have to set a `DataContext`, like you would do for any of the native controls, and you set your own through other means, for example the control's ctor or an Intializing function called before the properties are set.

In this situation, whether it is set or not, Avalonia will just overwrite it with the parent's `DataContext`.
Is there any way to avoid that at all? even forcing it back to a default context in `OnDataContextChanged` cause exception when it tries to cast the chid's DataContext to `MainViewModel` mindlessly.

Consider the following situation : 
![image](https://github.com/AvaloniaUI/Avalonia/assets/39861216/e185497d-6c2e-40dd-b3be-8bba98106a44)
![image](https://github.com/AvaloniaUI/Avalonia/assets/39861216/07df2d63-36c9-48cf-85a6-338772f53beb)

I need to provide a tileset, and 2 numerical value to this control.
Since no `DataContext` is provided, even if i set one in the ctor, it will still replace it with a `MainViewModel`
If i try to force it in `OnDataContextChanged` then the previously mentionned exception occurs on launch after a short while.

If i provide my own DataContext, then it will try to fetch it in that `DataContext` instead of the parent :
![image](https://github.com/AvaloniaUI/Avalonia/assets/39861216/c051616f-6e04-454f-b2c1-a1139c69e73b)

Any ideas on how i could get the result i want here?

**Comment by timunie** [comment]
If you want to set your data context in code, just don't bind to it...

**Reply by Mrcubix** [reply]
overwritten with `MainViewModel`

**Reply by timunie** [reply]
works for me: 

![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/eef29d17-3f8a-4371-8435-4276f69c3f38)

So I have no idea what you are doing different. I set the datacontext in contructor, just like you wish for a reason I can't understand (but that's okay)

**Reply by timunie** [reply]
will send another PR as for information, not meant to be merged.
